src/celeritas/phys/PrimaryGenerator.cc

```diff
@@ -0,0 +1,56 @@
+//----------------------------------*-C++-*----------------------------------//
+// Copyright 2020-2022 UT-Battelle, LLC, and other Celeritas developers.
+// See the top-level COPYRIGHT file for details.
+// SPDX-License-Identifier: (Apache-2.0 OR MIT)
+//---------------------------------------------------------------------------//
+//! \file celeritas/phys/PrimaryGenerator.cc
+//---------------------------------------------------------------------------//
+#include "PrimaryGenerator.hh"
+
+#include "corecel/cont/Range.hh"
+#include "celeritas/Types.hh"
+#include "celeritas/Units.hh"
+
+#include "PDGNumber.hh"
+
+namespace celeritas
+{
+//---------------------------------------------------------------------------//
+/*!
+ * Construct with options and shared particle data.
+ */
+PrimaryGenerator::PrimaryGenerator(SPConstParticles        particles,
+                                   PrimaryGeneratorOptions options)
+    : particles_(std::move(particles)), options_(options)
+{
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Generate primary particles.
+ */
+auto PrimaryGenerator::operator()() -> VecPrimary
+{
+    Primary p;
+    p.particle_id = particles_->find(PDGNumber{options_.pdg});
+    p.energy      = units::MevEnergy{options_.energy};
+    p.position    = options_.position;
+    p.direction   = options_.direction;
+    p.time        = 0;
```

Comment:
It's probably easier to store this baseline `Primary` as class data rather than the SP Particles and options.

Reply:
Good call!